### PR TITLE
Android signup

### PR DESCRIPTION
**Pull request by gcalero**
This PR aims to add this missing feature in android.

**Notes for reviewer**
The goal was reusing as much existing code as possible from AccountManager. Invoking accountManager->sendRequest requires to pass a callback object which is connected and whose methods are invoked. For this to work Qt requires that this object lives in a thread running a QEventLoop.
On the other hand AndroidHelper is a singleton whose instance is used - and created - in any thread, in general in android thread invoking native functions. For this reason we have to move AndroidHelper to a Qt thread, in this case the Qt main thread which runs the main looper.

**Test plan**
- Basic flow: a new High Fidelity account and log in with it
- Validation tests: try to force empty fields, invalid email address or username, short and long passwords
- User duplication: try to signup with an already registered email address or username
- Connectivity: try to signup using WIFI, LTE and with no connectivity

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13182/HighFidelity-Beta-PR14054-f4be767.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13182/HighFidelity-Beta-PR14054-f4be767.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13182/HighFidelity-Beta-Interface-PR14054-f4be767.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13182/HighFidelity-Beta-PR14054-f4be767.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13182/HighFidelity-Beta-Interface-PR14054-f4be767.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13356/HighFidelity-Beta-PR14054-6908107.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13356/HighFidelity-Beta-PR14054-6908107.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13356/HighFidelity-Beta-Interface-PR14054-6908107.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13356/HighFidelity-Beta-PR14054-6908107.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13356/HighFidelity-Beta-Interface-PR14054-6908107.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13392/HighFidelity-Beta-PR14054-fba3b48.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13392/HighFidelity-Beta-PR14054-fba3b48.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13392/HighFidelity-Beta-Interface-PR14054-fba3b48.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13392/HighFidelity-Beta-PR14054-fba3b48.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13392/HighFidelity-Beta-Interface-PR14054-fba3b48.exe

**Comment by amerhifi** [comment]
I'm noticing two items. When clicking on "forgot password" or "sign up" the hit area is very tiny and you have to be very precise.  Because you miss the hit box it acts like its laggy. 

Second item, When you go to the sign up or back from sign up it flashes between the ui.  Tested on S9+

**Comment by amerhifi** [comment]
Creating a new user with fake email example: abcabc@abc.com and some password, will allow a login without account being confirmed. Website will not allow the login

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13412/HighFidelity-Beta-PR14054-fba3b48.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13413/HighFidelity-Beta-PR14054-5fc24b2.apk

File: android/app/src/main/java/io/highfidelity/hifiinterface/fragment/SignupFragment.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+import io.highfidelity.hifiinterface.R;
+
+import static org.qtproject.qt5.android.QtActivityDelegate.ApplicationActive;
+import static org.qtproject.qt5.android.QtActivityDelegate.ApplicationInactive;
+
+public class SignupFragment extends Fragment {
+
+    private EditText mEmail;
+    private EditText mUsername;
+    private EditText mPassword;
```
Comment (amerhifi) [inline android/app/src/main/java/io/highfidelity/hifiinterface/fragment/SignupFragment.java]: how long is this fragment alive? Are we storing the password as raw text?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13413/HighFidelity-Beta-PR14054-5fc24b2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13413/HighFidelity-Beta-Interface-PR14054-5fc24b2.exe

File: interface/src/AndroidHelper.cpp [hunk 2]
```diff
@@ -67,3 +68,77 @@ void AndroidHelper::notifyHeadsetOn(bool pluggedIn) {
...
+    QJsonObject payload;
+
+    QJsonObject userObject;
+    userObject.insert("email", email);
+    userObject.insert("username", username);
+    userObject.insert("password", password);
+
+    payload.insert("user", userObject);
+
+    static const QString API_SIGNUP_PATH = "api/v1/users";
```
Comment (amerhifi) [inline interface/src/AndroidHelper.cpp]: we should have a way to store all configurations in one place. Possibly a static class with consts?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13413/HighFidelity-Beta-PR14054-5fc24b2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13413/HighFidelity-Beta-Interface-PR14054-5fc24b2.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13448/HighFidelity-Beta-PR14054-0d40d78.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13448/HighFidelity-Beta-PR14054-0d40d78.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13448/HighFidelity-Beta-Interface-PR14054-0d40d78.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13448/HighFidelity-Beta-PR14054-0d40d78.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13448/HighFidelity-Beta-Interface-PR14054-0d40d78.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13453/HighFidelity-Beta-PR14054-c0c95a1.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13453/HighFidelity-Beta-PR14054-c0c95a1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13453/HighFidelity-Beta-Interface-PR14054-c0c95a1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13453/HighFidelity-Beta-PR14054-c0c95a1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13453/HighFidelity-Beta-Interface-PR14054-c0c95a1.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13522/HighFidelity-Beta-PR14054-c0c95a1.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13523/HighFidelity-Beta-PR14054-b7c421b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13523/HighFidelity-Beta-PR14054-b7c421b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13523/HighFidelity-Beta-Interface-PR14054-b7c421b.dmg

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
Stopped testing; ran out of time before device had to be returned for end of day.

After creating an account, user is returned to the main page but is not logged in. I expected to be logged into the new account.

As expected, I could not duplicate existing emails or usernames. Invalid fields are not accepted.

However, leaving a username or password blank resulted in a message which sounds more like a failed login attempt for an existing account:  "Email, username or password incorrect."

Attempting to sign up with LTE connection resulted in a long wait followed by: "There was an unknown error while creating your account. Please try again later."

Attempting to sign up with no connection showed the same message immediately.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13538/HighFidelity-Beta-PR14054-1422c17.apk

**Comment by qaAdamSmith82** [comment]
I was taking over testing here, were there changes made?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13556/HighFidelity-Beta-PR14054-13e2b99.apk

**Comment by gcalero** [comment]
@ZappoMan @amerhifi @qaAdamSmith82 as other PRs were merged I had to push a couple of merge commits

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13556/HighFidelity-Beta-PR14054-13e2b99.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13556/HighFidelity-Beta-Interface-PR14054-13e2b99.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13556/HighFidelity-Beta-PR14054-13e2b99.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13556/HighFidelity-Beta-Interface-PR14054-13e2b99.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by conklin94122** [comment]
@gcalero Needs dev review (green check and label), can be merged if it passes.

**Pull request merged**